From abb53b2913425e7b7f897770539dcaa4cf0e9920 Mon Sep 17 00:00:00 2001
From: Aishwarya R <quic_aisr@quicinc.com>
Date: Tue, 9 Jul 2024 10:26:21 +0530
Subject: [PATCH] wifi: ath12k: Handle reboot when SSR recovery in progress

When SSR recovery in progress and reboot happens getting crash
as shutdown function proceed with core_deinit.
In reboot sequence, ahb_shutdown should return with a warning
that SSR recovery is interrupted due to shutdown request
when FLAG_RECOVERY is set and not proceed with core_deinit
as dependent modules might be unloaded already.

Patch-dependency: 710-001-ath12k-Add-AHB-driver-support-for-ipq5332.patch
Signed-off-by: Aishwarya R <quic_aisr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/ahb.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/ahb.c b/drivers/net/wireless/ath/ath12k/ahb.c
index d20ac54..335f9b4 100644
--- a/drivers/net/wireless/ath/ath12k/ahb.c
+++ b/drivers/net/wireless/ath/ath12k/ahb.c
@@ -1230,11 +1230,6 @@ err_core_free:
 
 static void ath12k_ahb_remove_prepare(struct ath12k_base *ab)
 {
-	if (test_bit(ATH12K_FLAG_RECOVERY, &ab->dev_flags)) {
-		ath12k_warn(ab, "SSR recovery in progress, interrupting ssr recovery due to shutdown request\n");
-		return;
-	}
-
 	set_bit(ATH12K_FLAG_UNREGISTERING, &ab->dev_flags);
 	set_bit(ATH12K_FLAG_UNREGISTERING, &ab->ag->dev_flags);
 	cancel_work_sync(&ab->restart_work);
@@ -1298,6 +1293,10 @@ static void ath12k_ahb_shutdown(struct platform_device *pdev)
 	 * remove() is invoked during rmmod & shutdown() during
 	 * system reboot/shutdown.
 	 */
+	if (test_bit(ATH12K_FLAG_RECOVERY, &ab->dev_flags)) {
+		ath12k_warn(ab, "SSR recovery in progress, interrupting ssr recovery due to shutdown request\n");
+		return;
+	}
 	ath12k_ahb_remove_prepare(ab);
 
 	if (!(test_bit(ATH12K_FLAG_REGISTERED, &ab->dev_flags)))
-- 
2.34.1

